README.md
Description
===========

* Installs OSEv3 and related packages.

Requirements
============

## Platform ##

* Tested on Red Hat RHEL 7.4
* Tested on Centos 7.4

## Openshift Version ##

* Support OSE version from 3.3+
* Support Origin version from 1.3+
* Default the installation to 3.7

**Highly recommended**: 

Explicitly set `node['is_apaas_openshift_cookbook']['ose_version']`, `node['is_apaas_openshift_cookbook']['ose_major_version']`
and ideally `node['is_apaas_openshift_cookbook']['docker_version']` to be safe when a major version is released on the
CentOS PaaS repository; 

This cookbook does support upgrade between major versions (Read doc)

## New Features!

- Control upgrades between versions
- ETCD migration v2 to v3 (Mandatory before upgrading to 3.7)

### Supported version
[x] 1.3 to 1.4

[x] 1.4 to 1.5

[x] 1.5 to 3.6

[x] 3.6 to 3.7

### Control Upgrade

Control upgrade requires variables to be set before running.

*The CHEF run will ignore any upgrade unless all variables are correctly set*

Variables:

| NAME | PURPOSE | Default value | Mandatory |
| ---------------- | ------------------------------- | ------------------ | ---------- |
| control_upgrade | Execute an upgrade     | `false`    | `YES` |
| control_upgrade_version | Target version (13,14,15,36,37)        |`""` |`YES`|
| control_upgrade_flag | Location of the control upgrade flag | `"/to_be_replaced"`  | `YES` |
| upgrade_repos | Target YUM repo | `""` | `NO` |

```json
  "override_attributes": {
    "is_apaas_openshift_cookbook": {
      "control_upgrade": true,
      "control_upgrade_version": "37",
      "control_upgrade_flag": "/tmp/ready",
      "upgrade_repos": [
        {
          "name": "centos-openshift-origin37",
          "baseurl": "http://mirror.centos.org/centos/7/paas/x86_64/openshift-origin37/",
          "gpgcheck": false
        }
      ]
    }
  }
```

[Read more about the upgrade process here!](docs/upgrade-process.md)

Test Matrix
===========

| Platform   | OSE 3.7.0 | OSE 3.6.1 | OSE 1.5.1 | OSE 1.4.1 | OSE 1.3.3 | OSE 1.2.2 |
| --------   | --------- | --------- | --------- | --------- | --------- | --------- |
| centos 7.4 | PASS      | PASS      | PASS      | PASS      | PASS      | Not supported |

Override Attributes
===================

[Read more about overriding attributes here!](docs/attribute-cookbook.md)

#### Structure ####
-------------------

* `node['is_apaas_openshift_cookbook']['openshift_master_identity_provider']['HTPasswdPasswordIdentityProvider']`

```json
{
"name" : "htpasswd_auth", 
"login" : true, 
"challenge" : true, 
"kind" : "HTPasswdPasswordIdentityProvider", 
"filename" : "/etc/openshift/openshift-passwd"
}
```

* `node['is_apaas_openshift_cookbook']['openshift_master_identity_provider']['LDAPPasswordIdentityProvider']`

```json
{
"name" : "ldap_identity", 
"login" : true, 
"challenge" : true, 
"kind" : "LDAPPasswordIdentityProvider", 
"ldap_server" : "ldap.domain.local",
 "ldap_bind_dn" : "", 
 "ldap_bind_password" : "",
 "ldap_insecure" : true, 
 "ldap_base_ou" : "OU=people,DC=domain,DC=local", 
 "ldap_preferred_username" : "uid"
}
```

* `node['is_apaas_openshift_cookbook']['openshift_master_asset_config']`

```json
{
 "extensionStylesheets":["/path/to/css"],
 "extensionScripts":["/path/to/script"],
 "extensions":"/path/to/my_images",
 "templates":"/path/to/template"
}
```

* `node['is_apaas_openshift_cookbook']['openshift_node_docker-storage']`

```json
{
 "DEVS": ["/dev/sdb","/dev/sdc"],
 "VG": "docker-vg",
 "DATA_SIZE": "80%FREE"
}
```

* `node['is_apaas_openshift_cookbook']['docker_log_driver']`

Set to `'json-file'` (default), `'journald'` or any other supported [docker log driver](https://docs.docker.com/engine/admin/logging/overview/).
Set to '' to disable it.

* `node['is_apaas_openshift_cookbook']['docker_log_options']`

Assuming `node['is_apaas_openshift_cookbook']['docker_log_driver']` is `'json-file'` (the default):

```json
{
 "max-size": "50M",
 "max-file": "3"
}
```

Any option can be set, as long as they are supported by the current [docker log driver](https://docs.docker.com/engine/admin/logging/overview/).


* `node['is_apaas_openshift_cookbook']['openshift_master_named_certificates']`

* CN or SAN names are automatically detected from the certificate file.

```json
[
  {
   "certfile": "/etc/where_is/my_certfile",
   "keyfile": "/etc/where_is/my_keyfile"
  }
]
```

* `node['is_apaas_openshift_cookbook']['persistent_storage']`

* They key called 'claim' is optional and will automatically create a PersistentVolumeClaim within a specified namespace.
 
```json
{
  "name": "registry",
  "capacity": "5Gi",
  "access_modes": "ReadWriteMany",
  "server": "core.domain.local",
  "path": "/var/nfs/registry",
  "claim": {
    "namespace": "default"
  }
}
```
* `node['is_apaas_openshift_cookbook']['openshift_hosted_cluster_metrics']`

Any option can be set, as long as they are supported by the current [Metrics deployer template](https://docs.openshift.com/container-platform/latest/install_config/cluster_metrics.html#metrics-ansible-variables).

Full list of attributes can be found [here](https://raw.githubusercontent.com/IshentRas/is_apaas_openshift_cookbook/master/attributes/metrics.rb).

We only support 1 cassandra POD.

Example of options for deploying metrics:
```json
{
  "openshift_hosted_cluster_metrics": true,
  "openshift_metrics_heapster_requests_memory": "1Gi",
  "openshift_metrics_image_version": "v1.5.1",
  "openshift_metrics_cassandra_storage_type": "dynamic",
  "openshift_metrics_hawkular_user_write_access": false,
  "....."
}
```
Example of removing metrics components:
```json
{
  "openshift_hosted_cluster_metrics": true,
  "openshift_metrics_install_metrics": false,
  "....."
}
```

Example of overriding the default kubelet options:
* `node['is_apaas_openshift_cookbook']['openshift_node_kubelet_args_custom']`
Any option can be set, as long as they are supported by current [Kubelet Options](https://kubernetes.io/docs/admin/kubelet/).

```json
{
  "openshift_node_kubelet_args_custom": {
    "pods-per-core": "5",
    "image-gc-high-threshold": "85",
    "resolv-conf": "/etc/resolv.conf"
  },
  "....."
}
```

Example of enabling a specific "feature-gate": enabling the [persistent local volumes feature](https://docs.openshift.com/container-platform/3.7/install_config/configuring_local.html#local-volume-enabling-local-volumes) (OSE v3.7 and up):
```json
{
  "openshift_master_api_server_args_custom": {
    "feature-gates": ["PersistentLocalVolumes=true"]
  },
  "openshift_master_controller_args_custom": {
    "feature-gates": ["PersistentLocalVolumes=true"]
  },
  "openshift_node_kubelet_args_custom": {
    "feature-gates": ["PersistentLocalVolumes=true"]
  }
}
```

Example of overriding the Audit Configuration Parameters for the `openshift_master_auditconfig` Hash

| Key Name   | Description |
| --------   | --------- |
| enable | A boolean to enable or disable audit logs. Default is `false`. |
| audit-file | File path where the requests should be logged to. If not set, logs are printed to master logs. |
| max-retention-day | Specifies maximum number of days to retain old audit log files based on the time stamp encoded in their filename. |
| max-file-number | Specifies the maximum number of old audit log files to retain. |
| max-file-size | Specifies maximum size in megabytes of the log file before it gets rotated. Defaults to 100MB. |

```json
{
  ...
  "openshift_master_auditconfig": {
    "enable": true,
    "audit-file": "/var/log/audit-ocp",
    "max-file-size": "50",
    "max-file-number": "10"
  },
  ...
}
```

Example of overriding the Audit Configuration Parameters for the `openshift_master_auditconfig` Hash

| Key Name   | Description |
| --------   | --------- |
| enable | A boolean to enable or disable audit logs. Default is `false`. |
| audit-file | File path where the requests should be logged to. If not set, logs are printed to master logs. |
| max-retention-day | Specifies maximum number of days to retain old audit log files based on the time stamp encoded in their filename. |
| max-file-number | Specifies the maximum number of old audit log files to retain. |
| max-file-size | Specifies maximum size in megabytes of the log file before it gets rotated. Defaults to 100MB. |

```json
{
  ...
  "openshift_master_auditconfig": {
    "enable": true,
    "audit-file": "/var/log/audit-ocp",
    "max-file-size": "50",
    "max-file-number": "10"
  },
  ...
}
```

Example of overriding the setting for global builds:
Any option can be set, as long as they are supported by current [Global Build Defaults and Overrides](https://docs.openshift.com/container-platform/latest/install_config/build_defaults_overrides.html).

**When setting CPU/Memory via `openshift_builddefaults_resources_*`, `openshift_builddefaults_resources` needs to be set to `true`**

```json
{
  "....."
  "openshift_builddefaults_git_http_proxy": "http://USER:PASSWORD@HOST:PORT",
  "openshift_builddefaults_git_https_proxy": "https://USER:PASSWORD@HOST:PORT",
  "openshift_builddefaults_git_no_proxy": "mycorp.com",
  "openshift_builddefaults_env": [
    {
      "http_proxy": "http://proxy.example.com.redhat.com:3128"
    },
    {
      "NO_PROXY": "ose3-master.example.com"
    }
  ],
  "openshift_builddefaults_image_labels": [
    {
      "name": "imagelabelname1",
      "value": "imagelabelvalue1"
    }
  ],
  "openshift_builddefaults_nodeselectors": {
    "nodelabel1": "nodelabelvalue1",
    "nodelabel2": "nodelabelvalue2"
  },
  "openshift_builddefaults_annotations": {
    "annotationkey1": "annotationvalue1"
  },
  "openshift_buildoverrides_force_pull": "true",
  "openshift_buildoverrides_image_labels": [
    {
      "name": "imagelabelname1",
      "value": "imagelabelvalue1"
    }
  ],
  "openshift_buildoverrides_nodeselectors": {
    "nodelabel1": "nodelabelvalue1"
  },
  "openshift_buildoverrides_annotations": {
    "annotationkey1": "annotationvalue1"
  },
  "....."
}
```

If OpenShift Container Platform administrators would like to control the level of overcommit and manage container density on nodes, masters can be configured to override the ratio between request and limit set on developer containers. This requires configuring the ClusterResourceOverride variables.

```json
{
  ".....",
  "openshift_clusterresourceoverrides_memorylimit_percent": "10",
  "openshift_clusterresourceoverrides_cpulimit_percent": "50",
  "openshift_clusterresourceoverrides_limitcputomem_percent": "200",
  "...."
}
```

## Cloud Providers Integration

Cloud providers integration requires passing some sensetive credentials to OpenShift. This cookbook uses encrypted data bags as the safest way to achieve this. Thus you should have: 

- running Chef Server
- encrypted data bags with cloud providers' credentials

*Currently only AWS integration is supported.*

### AWS 

To integrate your OpenShift installation with AWS you should have following attributes for `is_apaas_openshift_cookbook` cookbook:

```json
{
  "openshift_cloud_provider": "aws",
  "openshift_cloud_providers": {
    "aws": {
      "data_bag_name": "cloud-provider",
      "data_bag_item_name": "aws",
      "secret_file": "/etc/chef/shared_secret"
    }
  }
}
```

You should also have data bag named `cloud-provider` (`data_bag_name` attribute above) and encrypted with some shared secret data bag item named `aws` (`data_bag_item_name` attribute above) at your Chef Server. If `secret_file` attribute from above is *not* defined a default for Chef Client shared secret file will be used (`/etc/chef/encrypted_data_bag_secret`). For more information see [official Chef docs](https://docs.chef.io/data_bags.html#encrypt-a-data-bag-item).

Data bag item content should be of the form:

```json
{
  "id": "aws",
  "access_key_id": "your_access_key_id",
  "secret_access_key": "your_secret_access_key"
}
```

Please note: `id` value should be exactly the same as `data_bag_item_name` attribute value from above.

**Alternatively** you can attach IAM policies to your AWS instances and do *not* provide AWS credentials in encrypted data bags. In this case you should have the following attribute for `is_apaas_openshift_cookbook` cookbook:

```json
{
  "openshift_cloud_provider": "aws"
}
```

and the following IAM policy attached to your *master* servers:

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": "ec2:*",
      "Resource": "*"
    },
    {
      "Effect": "Allow",
      "Action": "elasticloadbalancing:*",
      "Resource": "*"
    }
  ]
}
```

and the following IAM policy attached to your *node* servers: 

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": "ec2:Describe*",
      "Resource": "*"
    },
    {
      "Effect": "Allow",
      "Action": "ec2:AttachVolume",
      "Resource": "*"
    },
    {
      "Effect": "Allow",
      "Action": "ec2:DetachVolume",
      "Resource": "*"
    }
  ]
}
```

=====

Include the default recipe in a CHEF role so as to ease the deployment. 

## Roles (Examples) 

* BASE

```json
{
  "name": "base",
  "description": "Common Base Role",
  "json_class": "Chef::Role",
  "default_attributes": {

  },
  "override_attributes": {
  },
  "chef_type": "role",
  "run_list": [
    "recipe[is_apaas_openshift_cookbook]"
  ],
  "env_run_lists": {

  }
}
```

* UNINSTALL (ADHOC)

```json
{
  "name": "uninstall",
  "description": "Common Base Role",
  "json_class": "Chef::Role",
  "default_attributes": {

  },
  "override_attributes": {
  },
  "chef_type": "role",
  "run_list": [
    "recipe[is_apaas_openshift_cookbook::adhoc_uninstall]"
  ],
  "env_run_lists": {

  }
}
```

* REDEPLOY CERTIFICATES (ADHOC)

The redeployment of the certificates requires variables and flags to be set before running.

*The CHEF run will ignore any attempts unless all variables are correctly set*

Variables:

| NAME | PURPOSE | Default value | Mandatory | Target |
| ---------------- | ------------------------------- | ------------------ | ---------- | ---------- |
| adhoc_redeploy_certificates | Execute certificates redeploy    | `false`    | `YES` | Masters/Nodes/Certificate Server |
| adhoc_redeploy_etcd_ca | Execute an ETCD CA redeploy    | `false`    | `YES` | ETCD/Masters |
| adhoc_redeploy_cluster_ca | Execute a CLUSTER CA CERTS redeploy   | `false` | `YES` | Masters/Nodes/Certificate Server |
| redeploy_etcd_ca_control_flag | Location of the etcd ca control flag | `"/to_be_replaced_ca_etcd"`  | `YES` | Certificate Server |

CHANGELOG.md
# Openshift 3 Cookbook CHANGELOG
This file is used to list changes made in each version of the Openshift 3 cookbook.

## v2.0.57
### Bug
- Fix issue when deleting servers

## v2.0.56
### Improvement
- Automatically upgrade the custom router template during upgrades

## v2.0.53-55
### Bug
- Fix issue with cloud_provider when restarting node service.

## v2.0.52
### Bug
- Fix issue reported 247 (Upgrade 36 -> 37)

## v2.0.51
### Bug
- Fix rubocop issues

## v2.0.50
### Bug
- Fix issue reported 244 (Insecure docker registry)
- Fix issue reported 247 (Customised CA)

### Improvement
- Give the possibility of providing an array for the SCCs

## v2.0.49
### Bug
- Fix several Cloning resource attributes issues
- Fix ETCD ca perms on certificate server

## v2.0.48
### Bug
- Prevent Docker to reinstall or update without control

## v2.0.47
### Improvement
- Asynchronous Errata Updates via variable asynchronous_upgrade (Default to false)

## v2.0.46
### Improvement
- Add possibility to specify specific yum options when installing docker docker_yum_options
- Add reverse lookup for SkyDNS

## v2.0.45
### Improvement
- Give the possibilty to provide a custom location for '/etc/NetworkManager/dispatcher.d/99-origin-dns.sh'

## v2.0.44
### Improvement
- Change the ETCD CA directory for the certificate server (Moving the certificate server becomes easier: mv /var/www/html)

## v2.0.43
### Bug
- Fix issue 240 when using standalone master

## v2.0.42
### Bug
- Regression bug for controller servers (TTL missing)

## v2.0.41
### Bug
- Remove the openshift-master.kubeconfig file when regenerating certs (Issue < 3.2)

## v2.0.40
### Bug
- Wrong option for generating node certs

## v2.0.39
### Bug
- Bug with certificates renewal for node servers

## v2.0.34 - v2.0.38
### Bug
- Bug with certificates renewal

## v2.0.33
### Bug
- Bug with library for identifying the server roles

## v2.0.32
### Bug
- Bug with library for identifying the server roles

## v2.0.31
### Bug
- Force secret for registry HTTPS

## v2.0.30
### Improvement
- Improve the certs renewal
- Improve the custom router template (Name can be customised) openshift_hosted_deploy_custom_name

## v2.0.29
### Improvement
- Stop pushing for etcd / docker package upgrade
- New vars for upgrade (ETCD/Docker) upgrade_docker_version upgrade_etcd_version

## v2.0.28
### Bug
- Fix issue when upgrading etcd

## v2.0.27
### Bug
- Fix foodcritic issue with new metric logic

## v2.0.26
### Bug
- Fix issue with excluder pkgs when server has got dual roles (master && node)
- Refactor the metrics logi to support all versions (3.3+)
- Fix issue when disabling SWAP for node servers

## v2.0.25
### Bug
- Add option to control docker yum exclude options custom_pkgs_excluder
- Revert last logic not working

## v2.0.24
### Bug
- Fix issue with master keys != 400 perms
### Improvement
- Add option to control docker yum options (Exclude/Include X pkgs) docker_yum_options

## v2.0.23
### Bug
- Fix issue with node service when NM is not turned off (Openshift >= 3.6)

## v2.0.22
### Bug
- Fix issue when restarting ETCD during upgrade

## v2.0.21
### Bug
- Fix issue during upgrade when having certificate_server which is not first master
- Add ETCD package to servers (Master for certificates, Etcd and certificate server for certificate renewal)

## v2.0.20
### Bug
- Fix issue during upgrade when upgrading master services

## v2.0.19
### Improvement
- Split the certificate server function from the other roles
- General code review
- Improve the general experience during certificate renewal

## v2.0.18
### Bug
- Fix issue during upgrade node components

## v2.0.17
### Improvement
- Give the possibility for setting ENV to router (openshift_hosted_deploy_env_router)

## v2.0.16
### Bug
- Fix issue during uninstall when server has not got docker installed

## v2.0.15
### Improvement
- Improve the general experience during upgrades
- Restarts of services are aligned with activities (upgrade/intall)
- Give the possibility for deploying custom router (openshift_hosted_deploy_custom_router)

## v2.0.14
### Bug
- Fix issue during upgrade when master are separated from ETCD

## v2.0.13
### Bug
- Fix race condition for certs
- Fix issue when deploying hosted services outside of the default project

## v2.0.12
### Bug
- General code improvements/MISC issues

## v2.0.11
### Bug
- General code improvements/MISC issues

## v2.0.10
### Bug
- Fix issue for upgrade when being a master only (No need to restart Docker)

## v2.0.9
### Bug
- Fix issue when creating PV (NFS)
- General code improvements/MISC issues

## v2.0.8
### Bug
- General code improvements/MISC issues

## v2.0.7
### Improvement 
- Improve creating PV/PVC

### Bug
- General code improvements/MISC issues

## v2.0.6
### Improvement 
- Add logic for removing redhat registry (Fixed issue 206)
- Improve IPtables logic

### Bug
- General code improvements/MISC issues

## v2.0.5
### Bug
- General code improvements/MISC issues

### Improvement 
- Initial Support for 3.7.x
- Capability for automated upgrade between versions
- Certitificate redeploy for ETCD (CA/CERTS)

## v1.10.67
### Bug
- Fixes error in v1.10.66 - use action ':nothing' on declaration, cf: https://serverfault.com/questions/587188/chef-how-to-run-a-resource-on-notification-only
- admin.kubeconfig made unambiguous - file creation failures avoided


## v1.10.66
### Improvement
- Add timeout to `du` call
- Rewrite ruby blocks into Chef resources

## v1.10.65
### Improvement
- Initial support for 3.6
- Capability for Overriding master and node servingInfo.minTLSVersion and .cipherSuites [openshift_(master|node)_cipher_suites, openshift_(master|node)_min_tls_version]
- Capability for defining ExternalIPNetworkCIDRs controls what values are acceptable for the service external IP field [openshift_master_external_ip_network_cidrs]
- Capability for defining IngressIPNetworkCIDR controls the range to assign ingress IPs [openshift_master_ingress_ip_network_cidr]

recipes/node.rb
#
# Cookbook Name:: is_apaas_openshift_cookbook
# Recipe:: node
#
# Copyright (c) 2015 The Authors, All Rights Reserved.

server_info = helper = OpenShiftHelper::NodeHelper.new(node)
node_servers = server_info.node_servers
certificate_server = server_info.certificate_server
is_node_server = server_info.on_node_server?
first_master = server_info.first_master
docker_version = node['is_apaas_openshift_cookbook']['openshift_docker_image_version']

ose_major_version = node['is_apaas_openshift_cookbook']['deploy_containerized'] == true ? node['is_apaas_openshift_cookbook']['openshift_docker_image_version'] : node['is_apaas_openshift_cookbook']['ose_major_version']
path_certificate = node['is_apaas_openshift_cookbook']['use_wildcard_nodes'] ? 'wildcard_nodes.tgz.enc' : "#{node['fqdn']}.tgz.enc"

if node['is_apaas_openshift_cookbook']['encrypted_file_password']['data_bag_name'] && node['is_apaas_openshift_cookbook']['encrypted_file_password']['data_bag_item_name']
  secret_file = node['is_apaas_openshift_cookbook']['encrypted_file_password']['secret_file'] || nil
  encrypted_file_password = data_bag_item(node['is_apaas_openshift_cookbook']['encrypted_file_password']['data_bag_name'], node['is_apaas_openshift_cookbook']['encrypted_file_password']['data_bag_item_name'], secret_file)
else
  encrypted_file_password = node['is_apaas_openshift_cookbook']['encrypted_file_password']['default']
end

if is_node_server
  ruby_block 'Turn off SWAP for nodes' do
    block do
      server_info.turn_off_swap
    end
    only_if { ::File.readlines('/etc/fstab').grep(/(^[^#].*swap.*)\n/).any? }
  end

  file '/usr/local/etc/.firewall_node_additional.txt' do
    content node['is_apaas_openshift_cookbook']['enabled_firewall_additional_rules_node'].join("\n")
    owner 'root'
    group 'root'
  end

  node['is_apaas_openshift_cookbook']['enabled_firewall_rules_node'].each do |rule|
    iptables_rule rule do
      action :enable
    end
  end

  directory node['is_apaas_openshift_cookbook']['openshift_node_config_dir'] do
    recursive true
  end

  if node['is_apaas_openshift_cookbook']['deploy_containerized']
    docker_image node['is_apaas_openshift_cookbook']['openshift_docker_node_image'] do
      tag docker_version
      action :pull_if_missing
    end

    docker_image node['is_apaas_openshift_cookbook']['openshift_docker_ovs_image'] do
      tag docker_version
      action :pull_if_missing
    end

    template '/etc/systemd/system/atomic-openshift-node-dep.service' do
      source 'service_node-deps-containerized.service.erb'
      notifies :run, 'execute[daemon-reload]', :immediately
    end

    template '/etc/systemd/system/atomic-openshift-node.service' do
      source 'service_node-containerized.service.erb'
      notifies :run, 'execute[daemon-reload]', :immediately
      variables(ose_major_version: ose_major_version)
    end

    template '/etc/systemd/system/openvswitch.service' do
      source 'service_openvswitch-containerized.service.erb'
      notifies :run, 'execute[daemon-reload]', :immediately
    end

    template '/etc/sysconfig/openvswitch' do
      source 'service_openvswitch.sysconfig.erb'
      notifies :restart, 'service[openvswitch]', :immediately unless node['is_apaas_openshift_cookbook']['upgrade']
    end
  else
    template '/etc/systemd/system/atomic-openshift-node.service' do
      source 'service_node.service.erb'
      notifies :run, 'execute[daemon-reload]', :immediately
      variables(ose_major_version: ose_major_version)
      only_if { ose_major_version.split('.')[1].to_i >= 6 }
    end
  end

  sysconfig_vars = {}

  if node['is_apaas_openshift_cookbook']['openshift_cloud_provider'] == 'aws'
    if node['is_apaas_openshift_cookbook']['openshift_cloud_providers']['aws']['data_bag_name'] && node['is_apaas_openshift_cookbook']['openshift_cloud_providers']['aws']['data_bag_item_name']
      secret_file = node['is_apaas_openshift_cookbook']['openshift_cloud_providers']['aws']['secret_file'] || nil
      aws_vars = data_bag_item(node['is_apaas_openshift_cookbook']['openshift_cloud_providers']['aws']['data_bag_name'], node['is_apaas_openshift_cookbook']['openshift_cloud_providers']['aws']['data_bag_item_name'], secret_file)

      sysconfig_vars['aws_access_key_id'] = aws_vars['access_key_id']
      sysconfig_vars['aws_secret_access_key'] = aws_vars['secret_access_key']
    end
  end

  template '/etc/sysconfig/atomic-openshift-node' do
    source 'service_node.sysconfig.erb'
    variables(sysconfig_vars)
    notifies :restart, 'service[Restart Node]', :immediately unless node['is_apaas_openshift_cookbook']['upgrade']
  end

  package 'atomic-openshift-node' do
    action :install
    version node['is_apaas_openshift_cookbook']['ose_version'] unless node['is_apaas_openshift_cookbook']['ose_version'].nil?
    options node['is_apaas_openshift_cookbook']['yum_options'] unless node['is_apaas_openshift_cookbook']['yum_options'].nil?
    not_if { node['is_apaas_openshift_cookbook']['deploy_containerized'] }
    retries 3
  end

  package 'atomic-openshift-sdn-ovs' do
    action :install
    version node['is_apaas_openshift_cookbook']['ose_version'] unless node['is_apaas_openshift_cookbook']['ose_version'].nil?
    options node['is_apaas_openshift_cookbook']['yum_options'] unless node['is_apaas_openshift_cookbook']['yum_options'].nil?
    only_if { node['is_apaas_openshift_cookbook']['openshift_common_use_openshift_sdn'] == true }
    not_if { node['is_apaas_openshift_cookbook']['deploy_containerized'] }
    retries 3
  end

  package 'conntrack-tools' do
    action :install
    not_if { node['is_apaas_openshift_cookbook']['deploy_containerized'] }
    retries 3
  end

  if node['is_apaas_openshift_cookbook']['adhoc_redeploy_cluster_ca']
    Chef::Log.warn("The CLUSTER CA CERTS redeploy will be skipped for Node[#{node['fqdn']}]. Could not find the flag: #{node['is_apaas_openshift_cookbook']['redeploy_cluster_ca_nodes_control_flag']}") unless ::File.file?(node['is_apaas_openshift_cookbook']['redeploy_cluster_ca_nodes_control_flag'])

    ruby_block "Redeploy CA certs for Node server: #{node['fqdn']}" do
      block do
        helper.remove_dir("#{node['is_apaas_openshift_cookbook']['openshift_node_config_dir']}/#{node['fqdn']}.tgz*")
      end
      only_if { ::File.file?(node['is_apaas_openshift_cookbook']['redeploy_cluster_ca_nodes_control_flag']) }
      notifies :delete, "file[#{node['is_apaas_openshift_cookbook']['redeploy_cluster_ca_nodes_control_flag']}]", :immediately
      notifies :restart, 'service[Restart Node]', :delayed if ::File.file?(node['is_apaas_openshift_cookbook']['redeploy_cluster_ca_nodes_control_flag'])
    end

    file node['is_apaas_openshift_cookbook']['redeploy_cluster_ca_nodes_control_flag'] do
      action :nothing
    end
  end

  remote_file "Retrieve certificate from Master[#{certificate_server['fqdn']}]" do
    path "#{node['is_apaas_openshift_cookbook']['openshift_node_config_dir']}/#{node['fqdn']}.tgz.enc"
    source "http://#{certificate_server['ipaddress']}:#{node['is_apaas_openshift_cookbook']['httpd_xfer_port']}/node/generated-configs/#{path_certificate}"
    action :create_if_missing
    notifies :run, 'execute[Un-encrypt node certificate tgz files]', :immediately
    notifies :run, 'execute[Extract certificate to Node folder]', :immediately
    retries 60
    retry_delay 5
  end

  execute 'Un-encrypt node certificate tgz files' do
    command "openssl enc -d -aes-256-cbc -in #{node['is_apaas_openshift_cookbook']['openshift_node_config_dir']}/#{node['fqdn']}.tgz.enc -out #{node['is_apaas_openshift_cookbook']['openshift_node_config_dir']}/#{node['fqdn']}.tgz -k '#{encrypted_file_password}'"
    action :nothing
  end

  execute 'Extract certificate to Node folder' do
    command "tar xzf #{node['fqdn']}.tgz && chown -R root:root ."
    cwd node['is_apaas_openshift_cookbook']['openshift_node_config_dir']
    action :nothing
  end

  directory "Fix permissions on #{node['is_apaas_openshift_cookbook']['openshift_node_config_dir']}" do
    path node['is_apaas_openshift_cookbook']['openshift_node_config_dir']
    owner 'root'
    group 'root'
    mode '0755'
  end

  file "Fix permissions on #{node['is_apaas_openshift_cookbook']['openshift_node_config_dir']}/ca.crt" do
    path ::File.join(node['is_apaas_openshift_cookbook']['openshift_node_config_dir'], 'ca.crt')
    owner 'root'
    group 'root'
    mode '0644'
  end

  remote_file '/etc/pki/ca-trust/source/anchors/openshift-ca.crt' do
    source "file://#{node['is_apaas_openshift_cookbook']['openshift_node_config_dir']}/ca.crt"
    notifies :run, 'ruby_block[Update ca trust]', :immediately
    sensitive true
  end

  # Use ruby_block for copying OpenShift CA to system CA trust
  ruby_block 'Update ca trust' do
    block do
      Mixlib::ShellOut.new('update-ca-trust').run_command
    end
    notifies :restart, 'service[docker]', :immediately if node['is_apaas_openshift_cookbook']['deploy_containerized']
    notifies :run, 'execute[Wait for 30 seconds for docker services to come up]', :immediately
    action :nothing
  end

  execute 'Wait for 30 seconds for docker services to come up' do
    command 'sleep 30'
    action :nothing
    only_if { node['is_apaas_openshift_cookbook']['deploy_containerized'] }
    not_if { node['is_apaas_openshift_cookbook']['upgrade'] }
  end

  if node['is_apaas_openshift_cookbook']['deploy_dnsmasq']
    package 'NetworkManager' do
      retries 3
    end

    template '/etc/origin/node/node-dnsmasq.conf' do
      source 'node-dnsmasq.conf.erb'
      only_if { ose_major_version.split('.')[1].to_i >= 6 }
    end

    template '/etc/dnsmasq.d/origin-dns.conf' do
      source 'origin-dns.conf.erb'
      variables(
        ose_major_version: ose_major_version
      )
      notifies :restart, 'service[dnsmasq]', :immediately
    end

    if node['is_apaas_openshift_cookbook']['custom_origin-dns']
      remote_file 'Retrieve custom file for 99-origin-dns.sh' do
        path '/etc/NetworkManager/dispatcher.d/99-origin-dns.sh'
        source "file://#{node['is_apaas_openshift_cookbook']['custom_origin_location']}"
        owner 'root'
        group 'root'
        mode '0755'
        notifies :restart, 'service[NetworkManager]', :immediately
      end
    else
      # On some systems, NetworkManager does not exist, so ignore_failure.
      cookbook_file '/etc/NetworkManager/dispatcher.d/99-origin-dns.sh' do
        source '99-origin-dns.sh'
        owner 'root'
        group 'root'
        mode '0755'
        action :create
        ignore_failure true
        notifies :restart, 'service[NetworkManager]', :immediately
      end
    end

    ruby_block 'Setup dnsmasq' do
      block do
        f = Chef::Util::FileEdit.new('/etc/dnsmasq.conf')
        f.insert_line_if_no_match(%r{^conf-dir=/etc/dnsmasq.d}, 'conf-dir=/etc/dnsmasq.d')
        f.write_file
      end
    end

    # ignore_failure in case this fails/is not necessary
    service 'dnsmasq' do
      action %i(enable start)
      ignore_failure true
    end

    ruby_block 'Enforce running NM_CONTROLLED on host (>= 3.6)' do
      block do
        f = Chef::Util::FileEdit.new("/etc/sysconfig/network-scripts/ifcfg-#{node['network']['default_interface']}")
        f.search_file_delete_line(/^NM_CONTROLLED/)
        f.write_file
      end
      notifies :restart, 'service[NetworkManager]', :immediately
      only_if { ::File.exist?("/etc/sysconfig/network-scripts/ifcfg-#{node['network']['default_interface']}") }
      only_if { ose_major_version.split('.')[1].to_i >= 6 }
    end
  end

  template node['is_apaas_openshift_cookbook']['openshift_node_config_file'] do
    source 'node.yaml.erb'
    variables(
      node_labels: node_servers.find { |server_node| server_node['fqdn'] == node['fqdn'] }['labels'].to_s.split(' '),
      ose_major_version: ose_major_version,
      kubelet_args: node['is_apaas_openshift_cookbook']['openshift_node_kubelet_args_default'].merge(node['is_apaas_openshift_cookbook']['openshift_node_kubelet_args_custom'])
    )
    notifies :run, 'execute[daemon-reload]', :immediately
    notifies :restart, 'service[Restart Node]', :immediately
    notifies :enable, 'systemd_unit[atomic-openshift-node]', :immediately
  end

  selinux_policy_boolean 'virt_use_nfs' do
    value true
  end

  execute 'Wait for API to become available before starting Node component' do
    command "[[ $(curl --silent ${MASTER_URL}/healthz/ready --cacert #{node['is_apaas_openshift_cookbook']['openshift_node_config_dir']}/ca.crt) =~ \"ok\" ]]"
    environment 'MASTER_URL' => node['is_apaas_openshift_cookbook']['openshift_HA'] ? node['is_apaas_openshift_cookbook']['openshift_master_api_url'] : "https://#{first_master['fqdn']}:#{node['is_apaas_openshift_cookbook']['openshift_master_api_port']}"
    retries 120
    retry_delay 1
    notifies :start, 'service[Restart Node]', :immediately unless node['is_apaas_openshift_cookbook']['upgrade'] && node['is_apaas_openshift_cookbook']['deploy_containerized']
    notifies :restart, 'service[Restart Node]', :immediately if node['is_apaas_openshift_cookbook']['upgrade'] && node['is_apaas_openshift_cookbook']['deploy_containerized']
  end
end
